File: sdk-java/src/test/java/io/superdurable/dex/iwfcompat/DeadEndFlow.java
/*
 * Copyright (c) 2026 Super Durable, Inc.
 *
 * Licensed under the Super Durable Source License 1.0.
 * You may not use this file except in compliance with the License.
 * See the LICENSE file in the repository root.
 *
 * SPDX-License-Identifier: LicenseRef-Super-Durable-1.0
 */

package io.superdurable.dex.iwfcompat;

import io.superdurable.dex.Channel;
import io.superdurable.dex.Context;
import io.superdurable.dex.Flow;
import io.superdurable.dex.PersistenceSchema;
import io.superdurable.dex.RPC;
import io.superdurable.dex.RPCResult;
import io.superdurable.dex.Step;
import io.superdurable.dex.StepDef;
import io.superdurable.dex.StepDecision;

import java.util.Collections;
import java.util.List;

final class DeadEndFlow implements Flow<Void> {
    final Channel<Void> idleSignal = Channel.define("idle-signal", Void.class);
    private final Step<Void> start = new Step<Void>() {
        @Override
        public Class<Void> getInputType() {
            return Void.class;
        }

        @Override
        public StepDecision execute(final Context context, final Void input) {
            return StepDecision.deadEnd();
        }
    };

    @Override
    public List<StepDef> getSteps() {
        return Collections.singletonList(StepDef.startStep(start));
    }

    @Override
    public PersistenceSchema getPersistenceSchema() {
        return PersistenceSchema.of(idleSignal);
    }

    @RPC
    public RPCResult<Integer> signalSize(final Context context) {
        return RPCResult.of(idleSignal.size(context));
    }

    @RPC
    public RPCResult<Integer> publishInternal(final Context context) {
        idleSignal.publish(context, null);
        return RPCResult.of(idleSignal.size(context));
    }
}

File: sdk-typescript/test/integ/dead_end_flow.ts
// Copyright (c) 2026 Super Durable, Inc.
//
// Licensed under the Super Durable Source License 1.0.
// You may not use this file except in compliance with the License.
// See the LICENSE file in the repository root.
//
// SPDX-License-Identifier: LicenseRef-Super-Durable-1.0

import {
  Channel,
  StepList,
  StepMovement,
  deadEnd,
  doubleCodec,
  gracefulComplete,
  rpc,
  stringCodec,
  voidCodec,
  type Context,
  type Flow,
  type PersistenceSchema,
  type RPCResult,
  type Step,
  type StepDecision,
} from "../../src/index.js";

class DeadEndStep implements Step<void> {
  public readonly inputCodec = voidCodec;

  public getStepType(): string {
    return "DeadEndStep";
  }

  public execute(_context: Context, _input: void): StepDecision {
    return deadEnd();
  }
}

class DeadEndCompleteStep implements Step<void> {
  public readonly inputCodec = voidCodec;

  public getStepType(): string {
    return "DeadEndCompleteStep";
  }

  public execute(_context: Context, _input: void): StepDecision {
    return gracefulComplete();
  }
}

export class DeadEndFlow implements Flow {
  public readonly idleSignal = new Channel("idle-signal", voidCodec);
  public readonly idleInternal = new Channel("idle-internal", voidCodec);
  private readonly start = new DeadEndStep();
  private readonly complete = new DeadEndCompleteStep();

  public getFlowType(): string {
    return "DeadEndFlow";
  }

  public getSteps() {
    return StepList.startStep(this.start).otherSteps(this.complete);
  }

  public getPersistenceSchema(): PersistenceSchema {
    return { channels: [this.idleSignal, this.idleInternal] };
  }

  @rpc({ outputCodec: doubleCodec })
  public signalSize(context: Context): RPCResult<number> {
    return { output: this.idleSignal.size(context) };
  }

  @rpc({ outputCodec: doubleCodec })
  public publishInternal(context: Context): RPCResult<number> {
    this.idleInternal.publish(context, undefined);
    return { output: this.idleInternal.size(context) };
  }

  @rpc({ inputCodec: stringCodec, outputCodec: doubleCodec })
  public invoke(context: Context, _input: string): RPCResult<number> {
    if (context.flowId === "" || context.runId === "") {
      throw new Error("invalid RPC context");
    }
    return { output: 100, nextSteps: [StepMovement.of(this.complete, undefined)] };
  }
}
